### ISSUE TITLE
[SDK 50] Dynamically overriding RTL settings causes the app to continuously reload

### ISSUE THREAD

**Issue by Subhi-Dev** [expo/expo#26532]
### Minimal reproducible example

https://snack.expo.dev/@subhi2006/rtl-loop-demo-run-in-sdk-50

### Summary

The previously working method (on SDK 49) of
```js
const shouldBeRTL = true;

if (shouldBeRTL !== I18nManager.isRTL && Platform.OS !== 'web') {
    I18nManager.allowRTL(shouldBeRTL);
    I18nManager.forceRTL(shouldBeRTL);
    Updates.reloadAsync();
}
```
That is specified in [Dynamically overriding RTL settings](https://docs.expo.dev/guides/localization/#dynamically-overriding-rtl-settings)
results in the app continuously reloading as shown when updating to SDK 50

![image](https://github.com/expo/expo/assets/53143122/db4bd214-0f2a-4120-8f03-e8c0de9738c8)

This only affects iOS
Note: Snack does not support SDK 50 and the demo above runs on SDK 49, the ```expo-updates``` library should be changed according to the version of the SDK in use ~~and it is possible that the error could be caused by the update to the library itself~~ the error is caused when updating to ```expo@50.0.0-alpha.0```

### Environment

  expo-env-info 1.2.0 environment info:
    System:
      OS: Windows 11 10.0.22621
    Binaries:
      Node: 21.4.0 - F:\nodejs\node.EXE
      npm: 10.2.0 - F:\nodejs\npm.CMD
    npmPackages:
      expo: ~50.0.2 => 50.0.2
      react: 18.2.0 => 18.2.0
      react-native: 0.73.2 => 0.73.2
    Expo Workflow: managed

**Comment by viown** [comment]
I can confirm that this happens to me as `I18nManager.allowRTL` and `I18nManager.forceRTL` don't seem to do anything on Expo Go as well as production builds.

**Comment by aleqsio** [comment]
Was able to repro, taking a look.

**Comment by dorhuri123** [comment]
Also facing this issue, is there a workaround in the meantime?

**Comment by viown** [comment]
> Also facing this issue, is there a workaround in the meantime?

You can downgrade to SDK 49 temporarily.

**Comment by dorhuri123** [comment]
if i am on development do i need to make another dev build again with the 49 SDK? because when i change the SDK i am getting an error

**Comment by aleqsio** [comment]
I think I traced this to problems with objc code in RN core returning invalid RTL for device locale – not sure if the real reason to blame is an iOS API change, still investigating

**Comment by aleqsio** [comment]
After looking into this a bit more I think it's working as expected once you follow the steps from:
[this Expo localization guide](https://docs.expo.dev/guides/localization/#enabling-rtl-support)

**Comment by Subhi-Dev** [comment]
@aleqsio I tried again and it still does the same problem. and this is the same guide I linked originally. Please make sure that your app is running with RTL forced and provide a repro if possible.

**Comment by aleqsio** [comment]
I've tested this in SDK50 with prebuild and the app builds and returns `18nManager.isRTL` true.

Here's a repro that works for me:
https://github.com/aleqsio/rtl_test

**Comment by topup-solutions** [comment]
Same issue, on Simulator and on real device opened via Expo

**Comment by Subhi-Dev** [comment]
Hello @aleqsio
I was able to build the app on iOS, however it still says "is RTL no"
So what I am trying to do is to force RTL on devices that have LTR languages as their main display language, my app doesn't have any English translations and I would like to show Arabic text correctly. 

The supportsRTL does in fact change the orientation to RTL but only if the user's device is RTL.

**Comment by MarouaniALA** [comment]
same issue on expo 49 on IOS `I18nManager.isRTL` returning false even with 
```
I18nManager.allowRTL(true);
I18nManager.forceRTL(true);
```

**Comment by MarouaniALA** [comment]
@Subhi-Dev did you find a solution please?

**Comment by Subhi-Dev** [comment]
@MarouaniALA No I have not yet, however if there is no solution you could try using ```isRTL``` with if statements as a temporary fix

**Comment by topup-solutions** [comment]
@MarouaniALA @Subhi-Dev 
This is from official documentation!
If you want to override the default RTL detection from your application code dynamically, you cannot use the static configuration in app config. Instead, apply these changes dynamically from your application code.

**This does not work in Expo Go, as Expo Go resets RTL preferences when opening the launcher or individual projects.**

But strange, I tested it with Android simulator and it works, unfortunately IOS not working. I think It one way to fix it use CSS with checking selected language and env settings of the  Device

**Comment by MarouaniALA** [comment]
@Subhi-Dev the problem is that I18nManager.isRTL only takes into account the language of the phone settings

**Comment by MarouaniALA** [comment]
@topup-solutions it works well on expo go the problem comes after building the app and even the android have a random strange behaviour when you change the layout font-awesome icons disappears unless you close the app and open it again

**Comment by Subhi-Dev** [comment]
Indeed @MarouaniALA You can use Async storage to store a variable isRTL and get it on every page it's needed, however treat this as a temporary solution as this will slow down your app.

**Comment by Subhi-Dev** [comment]
For a summary of the issue up to now
SDK 49:
Expo Go works
Custom builds stay on the system default and don't change to RTL when manually set

SDK 50:
Neither expo go nor custom builds work
They repeatedly reload on LTR devices

* This issue is only on iOS

**Comment by H4ck3r-x0** [comment]
facing the same issue! i did all the solutions mentioned above, did not work i installed expo-localization, and configred app.json
```
  "extra": {
      "supportsRTL": true
    },
    
      "plugins": [
      "expo-router",
      "expo-localization"
    ],
```

i did in the rootlayout

```
import { Redirect, Stack } from 'expo-router';
import { Text, I18nManager, TouchableOpacity } from 'react-native';
import { useSession } from '../../context/ctx';
import { useFonts } from 'expo-font';
import * as SplashScreen from 'expo-splash-screen';
import { useEffect } from 'react';
import colors from '../../constants/colors';
import { useRouter } from 'expo-router';
import { Feather } from '@expo/vector-icons';


SplashScreen.preventAutoHideAsync();
I18nManager.allowRTL(true);
I18nManager.forceRTL(false);

export default function AppLayout() {
    const router = useRouter();
    const [fontsLoaded, fontError] = useFonts({
        'T-b': require('../../assets/fonts/Tajawal-Bold.ttf'),
        'T-m': require('../../assets/fonts/Tajawal-Medium.ttf'),
        'T-r': require('../../assets/fonts/Tajawal-Regular.ttf'),
    });
    const { session, isLoading } = useSession();

    useEffect(() => {
        if (fontsLoaded || fontError) {
            SplashScreen.hideAsync();
        }
    }, [fontsLoaded, fontError]);

        console.log(I18nManager.isRTL)

    if (!fontsLoaded && !fontError) {
        return null;
    }

    if (isLoading) {
        return <Text>Loading...</Text>;
    }

    if (!session) {
        return <Redirect href="/sign-in" />;
    }

    return ( the rest of the component ... );
    
    ```
    
    i still get from         console.log(I18nManager.isRTL) = false!
    
    my layout is missed up.
    
    should i go back to sdk 49?

**Comment by H4ck3r-x0** [comment]
updates, when i build the app using eas, and install it on the Simulator, everything is fine, but when i published new update to eas and run the build with expo go, it is not working!

**Comment by H4ck3r-x0** [comment]
Hello @Subhi-Dev, okey I built the app using this command 
```
eas build -p ios --profile preview
```
Then I ran the build on the IOS Simulator, the RTL is working no problem event the Simulator language is in English. But when i published a new update to get the QR code to run the build with my own device wit Expo go,

```
eas update --branch preview --message "first build to test"
```
I ran the build with Expo go using my own device, but the RTL is missed up not working becuase my device language is in English, but if i switch the device language to Arabic my app on Expo Go RTL working as excepted.

Am not  using  development client.

I hope you got what i mean.

**Comment by Subhi-Dev** [comment]
So I did some research into the ```expo-localization``` package and found that removing the package allows RTL to be forced
This is probably a result of this section in the LocalizationModule.swift
```swift LocalizationModule.swift
  func setSupportsRTL(_ supportsRTL: Bool) {
    // These keys are used by React Native here: https://github.com/facebook/react-native/blob/main/React/Modules/RCTI18nUtil.m
    // We set them before React loads to ensure it gets rendered correctly the first time the app is opened.
    // On iOS we need to set both forceRTL and allowRTL so apps don't have to include localization strings.
    UserDefaults.standard.set(supportsRTL, forKey: "RCTI18nUtil_allowRTL")
    UserDefaults.standard.set(supportsRTL ? isRTLPreferredForCurrentLocale() : false, forKey: "RCTI18nUtil_forceRTL")
    UserDefaults.standard.synchronize()
  }
```
I am going to try to make a PR to solve this issue in the mean time **Removing the expo-localization plugin will fix this problem**

**Comment by aleqsio** [comment]
@Subhi-Dev You should be able to skip setting the `expo.extra.supportsURL` key in  `app.js` and that segment of code won't activate. You can still use the built-in react-native mechanism to control RTL dynamically.

**Comment by aleqsio** [comment]
I'm starting to think about adding a `forceRTL` field to the `extra` field that would force the app to run in RTL regardless of OS language. 

Would that be helpful for you, or do you need this to be dynamically configurable?

**Comment by Subhi-Dev** [comment]
Yes aleqsio that would be great indeed for applications that only support RTL which are plenty and would justify adding said variable.
However I think having the option to switch between RTL and LTR is essential for applications that switch between locales, so 
better documentation explaining what supportsRTL does and a modification of the Dynamically overriding RTL section would be much appreciated.

**Comment by Mahmoudkhoder01** [comment]
Hello guys, has anyone found a solution for this? on Android sometimes the RTLis works fine sometimes not for English phones, but for Arabic phones default language I8nManaget.isRtl returns true all the time even if I forced it to be false, and for ios the same but for both English and Arabic phones default language. Please anyone can help ASAP!

**Comment by Subhi-Dev** [comment]
@Mahmoudkhoder01 either you remove  the "supportsRTL": true from your app config or you delete the expo-localization library from your app

**Comment by SamJbori** [comment]
Facing the same issue on Expo 50 with `expo-localization` plugin configured as instructed, removing `"supportsRTL": true` didn't fix it, but removing `expo-localization` did

**Comment by Mahmoudkhoder01** [comment]
> Facing the same issue on Expo 50 with `expo-localization` plugin configured as instructed, removing `"supportsRTL": true` didn't fix it, but removing `expo-localization` did

But how do you let your app use the phone language?

**Comment by SamJbori** [comment]
> > Facing the same issue on Expo 50 with `expo-localization` plugin configured as instructed, removing `"supportsRTL": true` didn't fix it, but removing `expo-localization` did
> 
> But how do you let your app use the phone language?

Check my solution [here](https://stackoverflow.com/a/78206414/16082328)

**Comment by Mahmoudkhoder01** [comment]
Thank you all you saved my life!!

uninstall expo-localization worked for me

**Comment by billomore** [comment]
even removing it from the list of plugins works :) Thanks @Subhi-Dev!

**Comment by akarm13** [comment]
This is still an issue on Expo Go on SDK 50 and iPhone. It works fine on Expo 49. I wonder what has changed besides `expo-localization`. I've removed it from the plugins and even uninstalled it and it still causes infinite loop.

**Comment by ahmedamin-89** [comment]
Has anyone found a workaround?

**Comment by aleqsio** [comment]
@ahmedamin-89 If you need to use dynamic RTL mode (using `I18nManager.forceRTL`), move to development builds and prebuild, alternatively rely on the phone locale to choose RTL/LTR automatically – would that work for you?

**Comment by Mahmoudkhoder01** [comment]
> Has anyone found a workaround?

uninstall expo-localization package and remove it from the plugin in app.json, and use [this ](https://stackoverflow.com/questions/78197922/forcertl-on-ios-isnt-working-in-react-native-app/78206414#78206414)to get the lang of the device

**Comment by aleqsio** [comment]
@Mahmoudkhoder01 Why do you need to use `I18nManager.forceRTL` if the device should autoselect RTL for your application if it has the `supportsRTL` flag set?

**Comment by Asekon** [comment]
@aleqsio Thank you for your contributions towards this issue. until sdk 49 i was able to dynamically toggle between RTL and LTR on both ios and andoird, using expo go. Then I faced the issue mentioned here and waited for the fix. I am now on sdk 51   and the issue is fixed when i built the app for ios and tested it. However it works on expo go only when using android, which is very bothersome when testing. I know the docs state dynamically toggling isnt available on expo go but for some reason it always worked. Im using similar logic to the following. Always resets back to LTR after reloading on expo go ios.

``` 
 const changeLanguage = async () => {
    const rtl = I18nManager.isRTL;
    const lang = rtl ? "en" : "ar";
    i18n.changeLanguage(lang, () => {
      I18nManager.forceRTL(!rtl);
      Updates.reloadAsync();
    });
  };
  ```

**Comment by ahmedamin-89** [comment]
does anyone know how to make the language change from system settings on Apple devices?

**Comment by husam-m-97** [comment]
for me using expo with this implementaions bellow not changing rtl for the app 

  const onChangeLanguageHandler = async (id: string) => {
    if (currentLanguage === id) return;
    const isSelectedArabic = id === 'ar';
    i18n.changeLanguage(id);
    I18nManager.allowRTL(isSelectedArabic);
    I18nManager.forceRTL(isSelectedArabic);

    await storeObjectData('language', id);
    // Updates.reloadAsync();
  };

**Comment by manssorr** [comment]
Opps! after wasting whole day on this!

**Comment by aleqsio** [comment]
This issue should now be fixed on SDK51, but note that dynamically changing RTL/LTR in Expo Go will not work, as it would affect other projects.

You can now test RTL in expo go by using `forceRTL: true` in the same place as `supportsRTL`:

```
{
  "expo": {
    "extra": {
      "supportsRTL": true,
	  "forcesRTL": true
    },
    "plugins": ["expo-localization"]
  }
}
```

**Comment by XFAISAL** [comment]
> This issue should now be fixed on SDK51, but note that dynamically changing RTL/LTR in Expo Go will not work, as it would affect other projects.
> 
> You can now test RTL in expo go by using `forceRTL: true` in the same place as `supportsRTL`:
> 
> ```
> {
>   "expo": {
>     "extra": {
>       "supportsRTL": true,
> 	  "forcesRTL": true
>     },
>     "plugins": ["expo-localization"]
>   }
> }
> ```

For me still I18nManager.isRTL is return false

**Comment by ahmedamin-89** [comment]
> > This issue should now be fixed on SDK51, but note that dynamically changing RTL/LTR in Expo Go will not work, as it would affect other projects.
> > You can now test RTL in expo go by using `forceRTL: true` in the same place as `supportsRTL`:
> > ```
> > {
> >   "expo": {
> >     "extra": {
> >       "supportsRTL": true,
> > 	  "forcesRTL": true
> >     },
> >     "plugins": ["expo-localization"]
> >   }
> > }
> > ```
> 
> For me still I18nManager.isRTL is return false

Me too

**Comment by ahmedamin-89** [comment]
@XFAISAL did you find a fix?

**Comment by akarm13** [comment]
@ahmedamin-89

I fixed my issue by upgrading to Expo 51 and creating a development build. Expo Go should only be used for prototypes and that is for good [reason](https://docs.expo.dev/workflow/overview/#expo-go-an-optional-tool-for-learning). That is because every time you reload an app on Expo Go, it will reset its preferences such as I18nManager.isRTL. Development build actually keeps the preferences and it would work more like a production app and it is more reliable to compare to the final production version. If you are curious why it goes into a constant reload, take a look at this code:

```js
const shouldBeRTL = true;

if (shouldBeRTL !== I18nManager.isRTL) {
  I18nManager.allowRTL(shouldBeRTL);
  I18nManager.forceRTL(shouldBeRTL);
  Updates.reloadAsync();
}
```

`shouldBeRTL` is initially set to true for testing purposes. You should replace this with your own language selecting logic. In the app I am working on, I have three languages (English, Kurdish, Arabic). I simply check if the language is not English, then I know it is RTL. Since both Kurdish and Arabic are RTL languages.

Here is what happens if you use Expo Go, since it resets the RTL preferences on every reload based on the code above, let's say our app's language currently is in English (en)

1. `shouldBeRTL` is set to `true` for testing purposes.
2. Since the app's language is English, `I18nManager.isRTL` is initially `false`.
3. The code checks if `shouldBeRTL` (`true`) is not the same as `I18nManager.isRTL` (`false`).
4. The condition is met, so `I18nManager.allowRTL(true)` and `I18nManager.forceRTL(true)` are called.
5. `Updates.reloadAsync()` is called, causing the app to reload.
6. On reload, Expo Go resets `I18nManager.isRTL` back to `false` since, due to it being Expo Go, it reloads the preferences.
7. The condition `shouldBeRTL !== I18nManager.isRTL` is `true` again (`true !== false`), causing the code to run again.
8. This creates an infinite reload loop.

If you are worried about native code (editing Kotlin and Swift code manually), we have an internal app in our company with ~25k lines of code and I haven't touched the native code once. And if you are worried about constantly rebuilding the app from scratch after every change, it won't happen as often as you'd think. You only need it when you install a library that directly edits native code. I've rebuilt probably like 5-6 times and that's about it. (because of new Expo versions, new library versions that directly edit the native code, etc.) otherwise you won't rebuild it again.

**TL;DR:** Upgrade to Expo 51 and create a development build to avoid constant reload issues caused by Expo Go resetting preferences like `I18nManager.isRTL`. Development builds actually preserve these preferences, making it more reliable and similar to the final production version. Avoid using Expo Go beyond initial prototypes.

**Comment by Ahsanali012** [comment]
With  expo: "51.0.17" still the app wont toggle to RTL layout .

Tried following way did not work.

` useEffect(() => {
    I18nManager.allowRTL(true); // Enable RTL support
    I18nManager.forceRTL(true); // Set RTL layout direction based on isRTL value
   
  }, [lang, isRTL, dispatch]);`

**Comment by waleedAlmubarak** [comment]
This issue should be reopened. 

I'm using expo "51.0.21",

Setting `I18nManager.forceRTL(true);` doesn't work.
I must mention that `"forcesRTL": true` in expo extra app.json does work. But I need to change this dynamically so it will not be fixed soon app.json.

**Comment by Subhi-Dev** [comment]
Hello @waleedAlmubarak forcing RTL  dynamically seems to be working fine for me in SDK 51, perhaps you can send a repro of your problem.


I would also suggest following some of the advice provided in previous comments such as removing the expo-localization package and trying to use the development build instead of expo go.

**Comment by waleedAlmubarak** [comment]
@Subhi-Dev Removing `expo-localization` worked for me. 
Thank you for your feedback.

**Comment by JoelJohny** [comment]
Hello @aleqsio, forcing RTL dynamically was working fine for me in **SDK 51**. However, after updating to **SDK 52**, the app continuously reloads on Android, though it works in the development build. On **iOS**, forcing RTL works only after reloading twice, and most of the time, the alignment is incorrect (especially when the iOS device language is set to an RTL language like Arabic).

Will this issue be fixed soon?

**Comment by MrGhost23** [comment]
@aleqsio doesn't work well in SDK 52

**Comment by aleqsio** [comment]
Hi all,

I've tested this on SDK52 and it seems to work fine for me

<img width="362" alt="image" src="https://github.com/user-attachments/assets/62668c9a-d3e7-444d-a521-ecaa2d983d88" />

Locale is correctly reported as LTR but I18nManager returns true since we override it.

I tested with the `expo-localization` plugin both added to app.json and not added.

```
<Button
  title="toggle"
  onPress={() => {
    I18nManager.allowRTL(!I18nManager.isRTL);
    I18nManager.forceRTL(!I18nManager.isRTL);
    reloadAsync();
  }}
/>
```

**Comment by deepanshugoyal10** [comment]
Can we go from LTR to RTL without performing asyncReload? I tried using creating a provider wrapping my component and updating state every time i change my language, but it is not hard reloading

**Comment by Subhi-Dev** [comment]
Hey,
If you want to ask about a related issue or add to the issue here please do provide a Minimum Reproducible Project. It makes it at least 10 times easier to solve the issue and you will get an answer faster.
